### PR TITLE
Update tools.md for missing color meaning issue #1491

### PR DESCRIPTION
**Pull request by varshneydevansh**
## Issue > 
Null hypothesis test discrepancy #1491

## Solution Proposed >  
documentation needs to be enhanced

## Changes made >
Added the lines to the tools.md file which explains the following -

  1. Clarify the meaning of the colors in the output.
  2. Defining the 'success' and 'failure' explicitly.
  3. Clarify the interpretation of the p-value and the null hypothesis.

### CODE REVIEW

File: docs/tools.md [hunk 1]
```diff
@@ -186,6 +186,35 @@ Benchmark                               Time             CPU      Time Old
 This is a mix of the previous two modes, two (potentially different) benchmark binaries are run, and a different filter is applied to each one.
 As you can note, the values in `Time` and `CPU` columns are calculated as `(new - old) / |old|`.
 
+## Note: Interpreting the output
+
```
Comment (LebedevRI) [inline docs/tools.md:190]: This could use a snippet of `compare.py` output that contains said `pvalue` output.

File: docs/tools.md [hunk 2]
```diff
@@ -186,6 +186,35 @@ Benchmark                               Time             CPU      Time Old
...
+When comparing benchmarks, `compare.py` uses statistical tests to determine whether the performance has significantly changed.
```
Comment (LebedevRI) [inline docs/tools.md:191]: I think this should start with a more human-facing explanation first.
Something to the effect of:

```
Performance measurements are an art, and performance comparisons are doubly so.
Results are often noisy, and don't necessarily have large absolute difference between them,
so just by visual inspection, it is not at all apparent if two measurements are actually showing
a performance change or not. It is even more confusing with multiple benchmark repetitions.

Thankfully, what we can do, is use statistical tests on the results
to determine whether the performance has statistically-significantly changed.
`compare.py` uses [Mann–Whitney U test](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test),
with a null hypothesis being that there's no difference in performance.
```
and streamline the rest.

**Review by LebedevRI — COMMENTED** [review]
Thank you for looking into this!
I'm not sure how to best write this, but i do have some thoughts.

File: docs/tools.md [hunk 3]
```diff
@@ -186,6 +186,63 @@ Benchmark                               Time             CPU      Time Old
...
+  4. **Iterations:** The number of iterations the benchmark was run to get a stable estimate.
+
+  5. **Time Old and Time New:** These represent the average time it takes for a function to run in two different scenarios or versions. For example, you might be comparing how fast a function runs before and after you make some changes to it.
+
+  6. **CPU Old and CPU New:** These show the average amount of CPU time that the function uses in two different scenarios or versions. This is similar to Time Old and Time New, but focuses on CPU usage instead of overall time.
+
+
+In the comparison section, the relative differences in both time and CPU time are displayed for each input size.
+
+When comparing benchmarks, `compare.py` uses statistical tests to determine whether the performance has significantly changed.
```
Comment (LebedevRI) [inline docs/tools.md:219]:
```
When comparing benchmarks, `compare.py` uses statistical tests to determine whether there is a statistically-significant difference between the measurements being compared.
The result of said statistical test is additionally communicated through color coding:
```

File: docs/tools.md [hunk 4]
```diff
@@ -186,6 +186,63 @@ Benchmark                               Time             CPU      Time Old
...
+    'Success' means 'successfully finding a difference'.
+    'Failure' means 'failing to find a difference'.
+```
+A statistically significant difference is determined by a **p-value**, which is a measure of the probability that the observed difference could have occurred just by random chance. A smaller p-value indicates stronger evidence against the null hypothesis. 
+
+**Therefore:**
+  1. If the p-value is less than the chosen significance level (alpha), we reject the null hypothesis and conclude the benchmarks are significantly different.
+  2. If the p-value is greater than or equal to alpha, we fail to reject the null hypothesis and treat the two benchmarks as similar.
+
+
```
Comment (LebedevRI) [inline docs/tools.md:244]:
```
Also please note that **even if** we determine that there **is** a statistically-significant difference
between the two measurements, it does not _necessarily_ mean that the actual benchmarks
that were measured **are** different, or vice versa, even if we determine that
there is **no** statistically-significant difference between the two measurements,
it does not necessarily mean that the actual benchmarks that were measured **are not** different.
```

File: docs/tools.md [hunk 5]
```diff
@@ -186,6 +186,65 @@ Benchmark                               Time             CPU      Time Old
...
 
+## Note: Interpreting the output
+
+Performance measurements are an art, and performance comparisons are doubly so.
+Results are often noisy and don't necessarily have large absolute differences to them,
+so just by visual inspection, it is not at all apparent if two measurements are actually showing
+a performance change or not. It is even more confusing with multiple benchmark repetitions.
+
+Thankfully, what we can do, is use statistical tests on the results
+to determine whether the performance has statistically significantly changed.
```
Comment (LebedevRI) [inline docs/tools.md:197]: statistically-significantly

File: docs/tools.md [hunk 6]
```diff
@@ -186,6 +186,65 @@ Benchmark                               Time             CPU      Time Old
...
+  But remember, in this context:
+
+    'Success' means 'successfully finding a difference'.
+    'Failure' means 'failing to find a difference'.
+```
+A statistically significant difference is determined by a **p-value**, which is a measure of the probability that the observed difference could have occurred just by random chance. A smaller p-value indicates stronger evidence against the null hypothesis. 
+
+**Therefore:**
+  1. If the p-value is less than the chosen significance level (alpha), we reject the null hypothesis and conclude the benchmarks are significantly different.
+  2. If the p-value is greater than or equal to alpha, we fail to reject the null hypothesis and treat the two benchmarks as similar.
```
Comment (LebedevRI) [inline docs/tools.md:242]: I think this should go before the section that talks about the color coding.

**Review by varshneydevansh — COMMENTED** [review]
Thanks for your guidance, Roman. The starting paragraph felt so interesting while reading, and the changes you suggested made this more well versed.

It was the first time that I came across about benchmarking, learned so much in the process.

If anything, still felt inconsistent or incomplete, I would love to make those changes.

File: docs/tools.md [hunk 7]
```diff
@@ -186,6 +186,68 @@ Benchmark                               Time             CPU      Time Old
...
+
+
+A statistically-significant difference is determined by a **p-value**, which is a measure of the probability that the observed difference could have occurred just by random chance. A smaller p-value indicates stronger evidence against the null hypothesis. 
+
+**Therefore:**
+  1. If the p-value is less than the chosen significance level (alpha), we reject the null hypothesis and conclude the benchmarks are significantly different.
+  2. If the p-value is greater than or equal to alpha, we fail to reject the null hypothesis and treat the two benchmarks as similar.
+
+
+When comparing benchmarks, `compare.py` uses statistical tests to determine whether there is a statistically-significant difference between the measurements being compared.
```
Comment (LebedevRI) [inline docs/tools.md:226]: It seems like this sentence should not be here (just delete it)

File: docs/tools.md [hunk 8]
```diff
@@ -186,6 +186,68 @@ Benchmark                               Time             CPU      Time Old
...
+This might seem counter-intuitive if you're expecting 'green' to mean 'improved performance' and 'red' to mean 'worsened performance'. 
+```bash
+  But remember, in this context:
+
+    'Success' means 'successfully finding a difference'.
+    'Failure' means 'failing to find a difference'.
+```
+
+
+Also, please note that **even if** we determine that there **is** a statistically-significant difference between the two measurements, it does not _necessarily_ mean that the actual benchmarks that were measured **are** different, or vice versa, even if we determine that there is **no** statistically-significant difference between the two measurements, it does not necessarily mean that the actual benchmarks that were measured **are not** different.
```
Comment (LebedevRI) [inline docs/tools.md:247]: I believe, rest of the docs are linewrapped (w/ line width of 80),
so please format all the changed lines in this diff.

File: docs/tools.md (hunk 1 above)
Comment (LebedevRI) [reply docs/tools.md:190]: Eh, i think that could be useful, but it looks out of place there.
What i was thinking of was along the lines of `Example output` from above,
but one that does actually contain `pvalue`

File: docs/tools.md (hunk 7 above)
Comment (varshneydevansh) [reply docs/tools.md:226]: I deleted the sentence.

When comparing benchmarks, `compare.py` uses statistical tests to determine whether there is a statistically-significant difference between the measurements being compared.

**Comment by LebedevRI** [comment]
@dmah42 someone more fluent in english will need to proofread the wording,
i can only mostly help with the content here.

**Comment by dmah42** [comment]
> @dmah42 someone more fluent in english will need to proofread the wording, i can only mostly help with the content here.

seems good to me.  feel free to merge it whenever you're happy with the content.

File: docs/tools.md (hunk 1 above)
Comment (varshneydevansh) [reply docs/tools.md:190]: So, should I add an example like this? This is an hypothetical example.
```bash
Comparing BM_memcpy (from ./a.out) to BM_copy (from ./a.out)
Benchmark                               Time             CPU      Time Old      Time New       CPU Old       CPU New       P-value
--------------------------------------------------------------------------------------------------------------------------------------
[BM_memcpy vs. BM_copy]/8            +5.1649         +5.1637            37           230            37           230        0.001
[BM_memcpy vs. BM_copy]/64          +21.4352        +21.4374            74          1654            74          1653        0.0001
[BM_memcpy vs. BM_copy]/512        +143.6022       +143.5865            91         13122            91         13120        0.00001
```

Comment (LebedevRI) [reply docs/tools.md:190]: Let's add 
```
Benchmark                                                              Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------------
benchmark/threads:1/process_time/real_time_pvalue                    0.0000          0.0000      U Test, Repetitions: 27 vs 27
benchmark/threads:1/process_time/real_time_mean                     -0.1442         -0.1442            90            77            90            77
benchmark/threads:1/process_time/real_time_median                   -0.1444         -0.1444            90            77            90            77
benchmark/threads:1/process_time/real_time_stddev                   +0.3974         +0.3933             0             0             0             0
benchmark/threads:1/process_time/real_time_cv                       +0.6329         +0.6280             0             0             0             0
OVERALL_GEOMEAN                                                     -0.1442         -0.1442             0             0             0             0

```

Comment (varshneydevansh) [reply docs/tools.md:189]: Hi Roman,

I added the provided example in addition, I also added an explanation summary of each output line.

https://github.com/google/benchmark/pull/1624/commits/0f64e48fcd8bace2a26d918d4b5c99ebb8258224

File: docs/tools.md [hunk 9]
```diff
@@ -186,6 +186,146 @@ Benchmark                               Time             CPU      Time Old
 This is a mix of the previous two modes, two (potentially different) benchmark binaries are run, and a different filter is applied to each one.
 As you can note, the values in `Time` and `CPU` columns are calculated as `(new - old) / |old|`.
 
+
+4. The below output is a summary of a benchmark comparison with statistics
```
Comment (LebedevRI) [inline docs/tools.md:190]: It doesn't make sense to label this as `4.`, this is not a mode of operation.

File: docs/tools.md [hunk 10]
```diff
@@ -186,6 +186,146 @@ Benchmark                               Time             CPU      Time Old
...
+Results are often noisy and don't necessarily have large absolute differences to
+them, so just by visual inspection, it is not at all apparent if two
+measurements are actually showing a performance change or not. It is even more
+confusing with multiple benchmark repetitions.
+
+Thankfully, what we can do, is use statistical tests on the results to determine
+whether the performance has statistically-significantly changed. `compare.py`
+uses [Mann–Whitney U
+test](https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test), with a null
+hypothesis being that there's no difference in performance.
```
Comment (LebedevRI) [inline docs/tools.md:250]: Move this block to be before `4. The below output is a summary of a benchmark comparison with statistics`

**Review by LebedevRI — APPROVED** [review]
@varshneydevansh thank you!
I suppose this could use some more editorial work,
but this is certainly much better than what we have now :)

**Comment by varshneydevansh** [comment]
Thank you so much @LebedevRI for guiding me. I recently encountered benchmarking for the first time and began learning about it. Then I realized that this is something where, alongside learning, I can even contribute.

I learned a lot from this process, not just about benchmarking, but how to organize thoughts in a way that makes it easier to understand for others.

Regards.

**Pull request merged**